# BioE 131 Final Project: Multiple Sequence Analysis of Various SARS-CoV2 Sequences

Before we start we need to install and import all the necessary packages and libraries which we will use to run our analysis. These include Biopython, SeqIO, and MUSCLE. 

In [2]:
! pip install anaconda
! pip install biopython
! pip install muscle

  Created wheel for anaconda: filename=anaconda-0.0.1.1-py3-none-any.whl size=1122 sha256=aa924e7ef1f78445c6570895f3ae5c06da86e4748d5e1ee2cf96d679a70fce4a
  Stored in directory: /home/jovyan/.cache/pip/wheels/c2/f4/d5/04ee4841afe97419e952e4906651f4c55b39fb1038715b715e
Successfully built anaconda


In [2]:
from Bio import SeqIO

Once we have installed these different tools, we will use SeqIO to read the fasta files containing the genome sequences which we downloaded from GenBank. Each of these will be named according to the region from which they were extracted.

In [3]:
with open('wuhan_cov2', 'r') as f:
    china = list(SeqIO.parse(f, 'fasta'))
    
with open('mexico_cov2.fasta', 'r') as f:
    mexico = list(SeqIO.parse(f, 'fasta'))
    
with open('serbia_cov2.fasta', 'r') as f:
    serbia = list(SeqIO.parse(f, 'fasta'))
    
with open('maryland_cov2.fasta', 'r') as f:
    maryland = list(SeqIO.parse(f, 'fasta'))
    
with open('florida_cov2.fasta', 'r') as f:
    florida = list(SeqIO.parse(f, 'fasta'))

In order to run the MUSCLE software, we will first write a file called aeqs.fa that contains the 5 sequences we created before. This will allows us to run the alignment between them. 

In [4]:
with open('seqs.fa', 'w') as f:
    SeqIO.write(china, f, 'fasta')
    SeqIO.write(mexico, f, 'fasta')
    SeqIO.write(serbia, f, 'fasta')
    SeqIO.write(florida, f, 'fasta')
    SeqIO.write(maryland, f, 'fasta')

In [6]:
china[0].seq == mexico[0].seq

False

In [7]:
len(china[0].seq)

29903

In [8]:
! muscle -in seqs.fa -out seqs.aligned.fa


MUSCLE v3.8.1551 by Robert C. Edgar

http://www.drive5.com/muscle
This software is donated to the public domain.
Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.

seqs 5 seqs, lengths min 29782, max 29903, avg 29845
00:00:00     16 MB(1%)  Iter   1  100.00%  K-mer dist pass 1
00:00:00     16 MB(1%)  Iter   1  100.00%  K-mer dist pass 2
00:01:17  1051 MB(54%)  Iter   1  100.00%  Align node       
00:01:17  1051 MB(54%)  Iter   1  100.00%  Root alignment
00:02:14  1051 MB(54%)  Iter   2  100.00%  Refine tree   
00:02:14  1051 MB(54%)  Iter   2  100.00%  Root alignment
00:02:14  1051 MB(54%)  Iter   2  100.00%  Root alignment
00:04:39  1051 MB(54%)  Iter   3  100.00%  Refine biparts


## Analyzing gaps in the aligned genomes

In [7]:
"""
FUNCTION: seq_to_str
PARAMETER: sequence: a seq object from Seq.IO
RETURNS: the sequence as a Python string
"""
def seq_to_str(sequence):
    genome_sequence = []
    for base in sequence:
        genome_sequence.append(base)
    sequence = ''.join([base for base in sequence])
    return sequence

"""
FUNCTION: gaps
RETURNS: a dictionary --> keys: sequence.id; value: dictionary --> keys: nucleotide number, value: 1
IMPORTANCE: the dictionary gives an indication of where sequences had gaps (insertions/deletions) 
            when aligned to the consensus
"""
def gaps():
    sars_seq = []
    names = []
    substitutions = {}
    for sequence in SeqIO.parse("seqs.aligned.fa", "fasta"):
        sars_sequence = seq_to_str(sequence.seq)
        substitutions[sequence.id] = []
        for base in range(0, len(sequence.seq)):
            if sars_sequence[base] == '-':
                substitutions[sequence.id].append(base+1)
    return substitutions

"""
FUNCTION: results
PARAMATER: a dictionary --> keys: sequence.id; value: dictionary --> keys: nucleotide number, value: 1
RETURNS: the keys of the dictionary, which indicate what position of the genome had gaps
"""
def results(dictionary):
    for key in dictionary:
        print(key)
        print(dictionary[key])
        print()
        
number_of_substitutions = gaps()
results(number_of_substitutions)

MW310413.1
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35]

MW315981.1
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 29837, 29838, 29839, 29840, 29841, 29842, 29843, 29844, 29845, 29846, 29847, 29848, 29849, 29850, 29851, 29852, 29853, 29854, 29855, 29856, 29857, 29858, 29859, 29860, 29861, 29862, 29863, 29864, 29865, 29866, 29867, 29868, 29869, 29870, 29871, 29872, 29873, 29874, 29875, 29876, 29877, 29878, 29879, 29880, 29881, 29882, 29883, 29884, 29885, 29886, 29887, 29888, 29889, 29890, 29891, 29892, 29893, 29894, 29895, 29896, 29897, 29898, 29899, 29900, 29901, 29902, 29903]

MW309449.1
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 686, 687, 68

## Analysing the mutations in the sequences

In [10]:
# creating a dictionary of sequence.id --> sequence
sequence_dictionary = {}
for sequence in SeqIO.parse("seqs.aligned.fa", "fasta"):
    sars_sequence = seq_to_str(sequence.seq)
    sequence_dictionary[sequence.id] = sars_sequence

In [12]:
# creating a nested dictionary of sequence id --> {genome position --> nucleotide exchange}
align_to = sequence_dictionary['NC_045512.2']
mutations = {}
count = 0
for id, sequence in sequence_dictionary.items():
    mutations[id] = {}
    for base in range(0, len(align_to)):
        if sequence[base] != align_to[base] and sequence[base] !='-':
            mutations[id][base+1] = align_to[base] + '-->' + sequence[base]
            count +=1
mutations

{'MW310413.1': {241: 'C-->T',
  1578: 'T-->C',
  3037: 'C-->T',
  4226: 'C-->T',
  6618: 'G-->A',
  10741: 'C-->T',
  14408: 'C-->T',
  20268: 'A-->G',
  23403: 'A-->G',
  23604: 'C-->A',
  24076: 'T-->C',
  28854: 'C-->T',
  29266: 'G-->A',
  29710: 'T-->C',
  29862: 'G-->A',
  29864: 'G-->A',
  29868: 'G-->A',
  29870: 'C-->A',
  29894: 'A-->G'},
 'MW315981.1': {203: 'C-->T',
  241: 'C-->T',
  3037: 'C-->T',
  3095: 'T-->C',
  10265: 'G-->A',
  12880: 'C-->T',
  14408: 'C-->T',
  15972: 'A-->G',
  19718: 'C-->T',
  23403: 'A-->G',
  28077: 'G-->T',
  28759: 'T-->C',
  28881: 'G-->A',
  28882: 'G-->A',
  28883: 'G-->C',
  29692: 'G-->T'},
 'MW309449.1': {241: 'C-->T',
  376: 'G-->A',
  2416: 'C-->T',
  3037: 'C-->T',
  7471: 'C-->T',
  14408: 'C-->T',
  19862: 'C-->T',
  22899: 'G-->A',
  23403: 'A-->G',
  25563: 'G-->T',
  26233: 'G-->T',
  27881: 'C-->T',
  28250: 'T-->G'},
 'NC_045512.2': {},
 'MW315213.1': {42: 'T-->A',
  241: 'C-->T',
  3037: 'C-->T',
  3691: 'A-->C',
  10156: 'C

In [79]:
# creates a dictionary that indicates the frequency of mutations: genome_position --> frequency
mutation_frequency = {}
for id, seq_mutations in mutations.items():
    for base, mutation in seq_mutations.items():
        if base in mutation_frequency:
            mutation_frequency[base] += 1;
        else:
            mutation_frequency[base] = 1
# mutation_frequency

In [82]:
# creates a list of the most frequent mutations
most_frequent_mutations = []
for position, count in mutation_frequency.items():
    if count >= 2:
        most_frequent_mutations.append(position)
most_frequent_mutations

[241, 3037, 14408, 20268, 23403, 24076, 28854]

### Summary of mutations observed

| **reference genome position** 	| **nucleotide exchange** 	| **frequency** 	| **region**    	|
|:-----------------------------:	|:-----------------------:	|:-------------:	|---------------	|
|            **241**            	|       **C --> T**       	|    **4/4**    	|   **5' UTR**  	|
|            **3037**           	|       **C --> T**       	|    **4/4**    	|   **ORF1ab**  	|
|           **14408**           	|       **C --> T**       	|    **4/4**    	|   **ORF1ab**  	|
|           **20268**           	|       **A --> G**       	|    **2/4**    	|   **ORF1ab**  	|
|           **23403**           	|       **A --> G**       	|    **4/4**    	| **S protein** 	|
|           **24076**           	|       **T --> C**       	|    **2/4**    	| **S protein** 	|
|             25563             	|         G --> T         	|      1/4      	|     ORF3a     	|
|           **28854**           	|       **C --> T**       	|    **2/4**    	| **N protein** 	|
|             28881             	|         G --> A         	|      1/4      	|   N protein   	|
|             28882             	|         G --> A         	|      1/4      	|   N protein   	|
|             28883             	|         G --> C         	|      1/4      	|   N protien   	|